packages/framework-provider-azure/src/library/read-model-adapter.ts

```diff
@@ -18,7 +25,72 @@ export async function fetchReadModel(
     `[ReadModelAdapter#fetchReadModel] Loaded read model ${readModelName} with ID ${readModelID} with result:`,
     resource
   )
-  return [resource as ReadModelInterface]
+
+  if (!resource) return [resource as ReadModelInterface]
+
+  return [
+    {
+      ...resource,
+      boosterMetadata: {
+        ...resource?.boosterMetadata,
+        optimisticConcurrencyValue: resource._etag,
+      },
+    } as ReadModelInterface,
+  ]
+}
+
+async function insertReadModel(
+  logger: Logger,
+  readModel: ReadModelInterface,
+  db: CosmosClient,
+  config: BoosterConfig,
+  readModelName: string
+): Promise<void> {
+  const itemModel = {
+    ...readModel,
+    id: readModel.id.toString(),
+  } as ItemDefinition
+  try {
+    await db
+      .database(config.resourceNames.applicationStack)
+      .container(config.resourceNames.forReadModel(readModelName))
+      .items.create(itemModel)
+    logger.debug('[ReadModelAdapter#insertReadModel] Read model inserted')
+  } catch (err) {
+    // In case of conflict (The ID provided for a resource on a PUT or POST operation has been taken by an existing resource) we should retry it
+    if (err.code == 409) {
+      logger.debug('[ReadModelAdapter#insertReadModel] Read model insert failed with a conflict failure')
+      throw new OptimisticConcurrencyUnexpectedVersionError(err.message)
+    }
+    throw err
+  }
+}
+
+async function updateReadModel(
+  readModel: ReadModelInterface,
+  db: CosmosClient,
+  config: BoosterConfig,
+  readModelName: string,
+  logger: Logger
+): Promise<void> {
+  /** upsert only occurs if the etag we are sending matches the etag on the server. i.e. Only replace if the item hasn't changed */
+  const options = {
+    accessCondition: { condition: readModel.boosterMetadata?.optimisticConcurrencyValue || '*', type: 'IfMatch' },
+  } as RequestOptions
+  try {
+    await db
+      .database(config.resourceNames.applicationStack)
+      .container(config.resourceNames.forReadModel(readModelName))
+      .items.upsert(readModel, options)
+    logger.debug('[ReadModelAdapter#updateReadModel] Read model updated')
+  } catch (err) {
+    // If there is a precondition failure then we should retry it
+    if (err.code == 412) {
```

Comment:
Same here, maybe...

```suggestion
    if (err.code == AZURE_PRECONDITION_FAILED_ERROR) {
```

Maybe there's some Azure library that provides an easier to read enum for these numbers?